File: scripts/build_kb.py
"""
Build RAG knowledge base from PDFs using PyMuPDF for text extraction.
Handles PDFs with embedded text (including those pdf-parse fails on).

Usage:
    python scripts/build_kb.py
"""

import json, re, sys, time
from pathlib import Path

PDF_DIR = Path(__file__).resolve().parent.parent / 'book_for_rag'
OUTPUT = Path(__file__).resolve().parent.parent / 'data' / 'kb' / 'combined_knowledge.json'

CHUNK_SIZE = 800
CHUNK_OVERLAP = 150
MIN_CHUNK_CHARS = 60

def clean_text(raw: str) -> str:
    return re.sub(r'[\x00-\x08\x0b\x0c\x0e-\x1f]', '',
           re.sub(r'\r\n|\f', '\n',
           re.sub(r'[ \t]+', ' ', raw))).strip()

def chunk_text(text: str, page_num: int, source_path: str, doc_id: str, idx_ref: list) -> list:
    """Split text into chunks using sentence/paragraph boundaries."""
    if not text or not text.strip():
        return []
    plain = text.strip()

    # Split into segments by paragraph breaks, then sentence endings
    separators = ['\n\n', '。', '！', '？', '\n']
    segments = [plain]
    for sep in separators:
        next_seg = []
        for seg in segments:
            if len(seg) <= CHUNK_SIZE * 1.5:
                next_seg.append(seg)
            else:
                parts = [p for p in seg.split(sep) if p.strip()]
                next_seg.extend(parts)
        segments = next_seg
        if all(len(s) <= CHUNK_SIZE * 1.2 for s in segments):
            break

    # Merge segments into chunks
    chunks = []
    current = ''
    for seg in segments:
        seg = seg.strip()
        if not seg:
            continue
        candidate = current + '。' + seg if current else seg
        if len(candidate) <= CHUNK_SIZE:
            current = candidate
        else:
            if current and len(current) >= MIN_CHUNK_CHARS:
                idx_ref[0] += 1
                chunks.append({
                    'chunk_id': f'{doc_id}-p{page_num:04d}-c{idx_ref[0]:04d}',
                    'document_id': doc_id,
                    'title': Path(source_path).stem,
                    'source_path': str(source_path),
                    'page_start': page_num,
                    'page_end': page_num,
                    'chunk_index': idx_ref[0],
                    'char_count': len(current),
                    'text': current,
                    'text_preview': current[:200]
                })
            # Keep overlap from tail
            overlap = current[-CHUNK_OVERLAP:] if len(current) > CHUNK_OVERLAP else current
            current = overlap + '。' + seg
            if len(current) > CHUNK_SIZE * 2:
                current = current[-CHUNK_SIZE:]

    if current and len(current.strip()) >= MIN_CHUNK_CHARS:
        idx_ref[0] += 1
        current = current.strip()
        chunks.append({
            'chunk_id': f'{doc_id}-p{page_num:04d}-c{idx_ref[0]:04d}',
            'document_id': doc_id,
            'title': Path(source_path).stem,
            'source_path': str(source_path),
            'page_start': page_num,
            'page_end': page_num,
            'chunk_index': idx_ref[0],
            'char_count': len(current),
            'text': current,
            'text_preview': current[:200]
        })
    return chunks

def process_pdf(file_path: Path):
    """Extract text from a PDF using PyMuPDF."""
    import pymupdf
    doc = pymupdf.open(str(file_path))
    pages_text = []
    for i in range(len(doc)):
        text = doc[i].get_text().strip()
        pages_text.append(text)
    total = len(doc)
    doc.close()
    return {
        'numPages': total,
        'pages': pages_text,
        'text': '\n\n'.join(pages_text)
    }

def main():
    print('Building knowledge base from PDFs...\n')

    pdf_files = sorted(PDF_DIR.glob('*.pdf'))
    if not pdf_files:
        print('No PDFs found!')
        sys.exit(1)

    print(f'Found {len(pdf_files)} PDFs:\n' +
          '\n'.join(f'  - {p.name}' for p in pdf_files) + '\n')

    all_chunks = []
    total_pages = 0
    doc_id = 'mental-health-zh-v1'
    idx_ref = [0]
    processed = 0
    skipped = []

    for pdf_path in pdf_files:
        name = pdf_path.name
        print(f'  [{processed+1}/{len(pdf_files)}] {name} ... ', end='', flush=True)

        try:
            result = process_pdf(pdf_path)
            file_chunks = []

            for i, page_text in enumerate(result['pages']):
                cleaned = clean_text(page_text)
                if not cleaned:
                    continue
                page_chunks = chunk_text(cleaned, i + 1, str(pdf_path), doc_id, idx_ref)
                file_chunks.extend(page_chunks)

            all_chunks.extend(file_chunks)
            total_pages += result['numPages']

            if file_chunks:
                print(f'{len(file_chunks)} chunks / {result["numPages"]} pages')
                processed += 1
            else:
                print(f'0 chunks (scanned/no text)')
                skipped.append(name)

        except Exception as e:
            print(f'ERROR: {e}')
            skipped.append(name)

    # Build final JSON
    kb = {
        'document': {
            'document_id': doc_id,
            'title': '心理健康中文综合知识库（指南 + 课本）',
            'source_path': str(PDF_DIR),
            'generated_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
            'parser_backend': 'PyMuPDF',
            'page_count': total_pages,
            'chunk_count': len(all_chunks),
            'chunk_size': CHUNK_SIZE,
            'chunk_overlap': CHUNK_OVERLAP,
            'min_chunk_size': MIN_CHUNK_CHARS,
            'tags': ['mental-health', 'Chinese-guidelines', 'psychiatry', 'psychology']
        },
        'chunks': all_chunks
    }

    OUTPUT.parent.mkdir(parents=True, exist_ok=True)
    with open(OUTPUT, 'w', encoding='utf-8') as f:
        json.dump(kb, f, ensure_ascii=False, indent=2)

    size_mb = OUTPUT.stat().st_size / 1024 / 1024
    print(f'\nDone!')
    print(f'   Processed: {processed} PDFs')
    print(f'   Skipped (scanned): {len(skipped)}')
    if skipped:
        print(f'   - ' + '\n   - '.join(skipped))
    print(f'   Total pages: {total_pages}')
    print(f'   Total chunks: {len(all_chunks)}')
    print(f'   Output: ~{size_mb:.1f} MB')

if __name__ == '__main__':
    main()

File: scripts/build_kb.js
const fs = require('fs');
const path = require('path');
const PDF_DIR = path.resolve(__dirname, '..', 'book_for_rag');
const OUTPUT = path.resolve(__dirname, '..', 'data', 'kb', 'combined_knowledge.json');

const CHUNK_SIZE = 800;
const CHUNK_OVERLAP = 150;
const MIN_CHUNK_CHARS = 60;

// Simple hash for chunk IDs
const simpleHash = (text) => {
    let h = 0;
    for (let i = 0; i < text.length; i++) {
        h = ((h << 5) - h) + text.charCodeAt(i);
        h |= 0;
    }
    return (h >>> 0).toString(16).padStart(8, '0');
};

// Clean text
function cleanText(raw) {
    return raw
        .replace(/\r\n/g, '\n')
        .replace(/\f/g, '\n')
        .replace(/[ \t]+/g, ' ')
        .replace(/\n{4,}/g, '\n\n\n')
        .replace(/[\x00-\x08\x0b\x0c\x0e-\x1f]/g, '')
        .trim();
}

// Split text into chunks with character-based overlap
function chunkText(text, pageNum, sourcePath, docId, chunkIndexRef) {
    const chunks = [];
    const plain = text.trim();
    if (!plain) return chunks;

    // Approach: split by sentences/paragraphs for coherence, then merge
    const separators = ['\n\n', '。', '！', '？', '\n'];
    let segments = [plain];

    for (const sep of separators) {
        const next = [];
        for (const seg of segments) {
            if (seg.length <= CHUNK_SIZE * 1.5) {
                next.push(seg);
            } else {
                next.push(...seg.split(sep).filter(s => s.trim()));
            }
        }
        segments = next;
        if (segments.every(s => s.length <= CHUNK_SIZE * 1.2)) break;
    }

    // Merge segments into chunks of ~CHUNK_SIZE
    let current = '';
    for (const seg of segments) {
        const trimmed = seg.trim();
        if (!trimmed) continue;
        const candidate = current ? current + '。' + trimmed : trimmed;

        if (candidate.length <= CHUNK_SIZE) {
            current = candidate;
        } else {
            // Emit current, keep overlap from tail
            if (current.length >= MIN_CHUNK_CHARS) {
                chunkIndexRef.val++;
                chunks.push({
                    chunk_id: `${docId}-p${String(pageNum).padStart(4, '0')}-c${String(chunkIndexRef.val).padStart(4, '0')}`,
                    document_id: docId,
                    title: path.basename(sourcePath).replace('.pdf', ''),
                    source_path: sourcePath,
                    page_start: pageNum,
                    page_end: pageNum,
                    chunk_index: chunkIndexRef.val,
                    char_count: current.length,
                    text: current,
                    text_preview: current.slice(0, 200)
                });
                // Keep trailing CHUNK_OVERLAP chars as overlap
                current = current.length > CHUNK_OVERLAP
                    ? current.slice(-CHUNK_OVERLAP) + '。' + trimmed
                    : current + '。' + trimmed;
            } else {
                current = candidate;
            }
        }
    }

    if (current && current.trim().length >= MIN_CHUNK_CHARS) {
        chunkIndexRef.val++;
        const trimmed = current.trim();
        chunks.push({
            chunk_id: `${docId}-p${String(pageNum).padStart(4, '0')}-c${String(chunkIndexRef.val).padStart(4, '0')}`,
            document_id: docId,
            title: path.basename(sourcePath).replace('.pdf', ''),
            source_path: sourcePath,
            page_start: pageNum,
            page_end: pageNum,
            chunk_index: chunkIndexRef.val,
            char_count: trimmed.length,
            text: trimmed,
            text_preview: trimmed.slice(0, 200)
        });
    }

    return chunks;
}

async function processPDF(filePath) {
    const buf = fs.readFileSync(filePath);
    const { PDFParse } = await import('pdf-parse');
    const pdf = new PDFParse(new Uint8Array(buf));
    await pdf.load();
    const result = await pdf.getText();

    // result = { pages: [{text: "..."}], text: "all text", total: numPages }
    return {
        numPages: result.total || result.pages?.length || 1,
        pages: result.pages ? result.pages.map(p => p.text || '') : [],
        text: result.text || ''
    };
}

async function main() {
    console.log('Scanning PDFs in book_for_rag/...\n');

    const files = fs.readdirSync(PDF_DIR)
        .filter(f => f.toLowerCase().endsWith('.pdf'))
        .sort()
        .map(f => path.join(PDF_DIR, f));

    if (!files.length) {
        console.error('No PDF files found in book_for_rag/');
        process.exit(1);
    }

    console.log(`Found ${files.length} PDFs:\n${files.map(f => '  - ' + path.basename(f)).join('\n')}\n`);

    const allChunks = [];
    let totalPages = 0;
    const docId = 'mental-health-zh-v1';
    const chunkIndexRef = { val: 0 };

    for (const filePath of files) {
        const fileName = path.basename(filePath);
        process.stdout.write(`  Processing: ${fileName} ... `);

        try {
            const result = await processPDF(filePath);
            const fileChunks = [];

            if (result.pages.length > 1) {
                for (let i = 0; i < result.pages.length; i++) {
                    const cleaned = cleanText(result.pages[i]);
                    if (!cleaned) continue;
                    const pageChunks = chunkText(cleaned, i + 1, filePath, docId, chunkIndexRef);
                    fileChunks.push(...pageChunks);
                }
            } else if (result.text && result.text.trim()) {
                const cleaned = cleanText(result.text);
                if (cleaned) {
                    const pageChunks = chunkText(cleaned, 1, filePath, docId, chunkIndexRef);
                    fileChunks.push(...pageChunks);
                }
            }

            allChunks.push(...fileChunks);
            totalPages += result.numPages;
            console.log(`done - ${fileChunks.length} chunks from ${result.numPages} pages`);
        } catch (err) {
            console.log(`ERROR: ${err.message}`);
        }
    }

    // Build final JSON
    const knowledgeBase = {
        document: {
            document_id: docId,
            title: '心理健康中文综合知识库（指南 + 课本）',
            source_path: PDF_DIR,
            generated_at: new Date().toISOString(),
            parser_backend: 'pdf-parse v2 (pdf.js)',
            page_count: totalPages,
            chunk_count: allChunks.length,
            chunk_size: CHUNK_SIZE,
            chunk_overlap: CHUNK_OVERLAP,
            min_chunk_size: MIN_CHUNK_CHARS,
            tags: ['mental-health', 'Chinese-guidelines', 'psychiatry', 'psychology']
        },
        chunks: allChunks
    };

    fs.mkdirSync(path.dirname(OUTPUT), { recursive: true });
    fs.writeFileSync(OUTPUT, JSON.stringify(knowledgeBase, null, 2), 'utf-8');

    console.log(`\nDone!`);
    console.log(`   Total PDFs: ${files.length}`);
    console.log(`   Total pages: ${totalPages}`);
    console.log(`   Total chunks: ${allChunks.length}`);
    console.log(`   Output: ${OUTPUT}`);
    console.log(`   File size: ${(fs.statSync(OUTPUT).size / 1024 / 1024).toFixed(1)} MB`);
}

main().catch(err => {
    console.error('Fatal error:', err);
    process.exit(1);
});
